### PR TITLE
[FIX] Partial evaluator only uses fma for elementwise multiplication

### PR DESCRIPTION
**Pull request by SteveBronder**
This PR does a check during the partial evaluator when it sees expressions such as `x * y + z` and only does `fma` if `x` and `y` are not both matrix types. It also adds another pattern match for `Add [z; EltTimes__ [x; y]]` since those are ones we do want to transform to use `fma(x, y, z)`. 

On a side note, I wonder if we should have like a `MatrixTimes__` to separate out the difference between scalar and matrix multiplication?

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [x] OR, no user-facing changes were made

## Release notes

Fixes https://github.com/stan-dev/cmdstan/issues/1077 so partial evaluator only uses fma for elementwise multiplication

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: src/analysis_and_optimization/Partial_evaluator.ml [hunk 1]
```diff
@@ -787,11 +787,41 @@ let rec eval_expr ?(preserve_stability = false) (e : Expr.Typed.t) =
...
                   , [ z
                     ; { pattern=
                           FunApp (StanLib ("Times__", FnPlain, mem), [x; y])
                       ; _ } ] )
+                  when (not preserve_stability)
+                       && not
+                            ( UnsizedType.is_eigen_type x.meta.type_
+                            && UnsizedType.is_eigen_type y.meta.type_ ) ->
+                    let lub_mem = lub_mem_pat [mem] in
+                    FunApp (StanLib ("fma", suffix, lub_mem), [x; y; z])
```
Comment (WardBrian) [inline src/analysis_and_optimization/Partial_evaluator.ml:806]: These first two cases can stay together, right?

Comment (SteveBronder) [reply src/analysis_and_optimization/Partial_evaluator.ml:806]: They actually have to be broken up, at least when I did this I get a 

```
Error (warning 57): Ambiguous or-pattern variables under guard;
variables x,y may match different arguments. (See manual section 9.5) ocamllsp
```

Comment (WardBrian) [reply src/analysis_and_optimization/Partial_evaluator.ml:806]: Huh, today I learned. There's a great explanation of why this is true in the manual: https://ocaml.org/releases/4.12/htmlman/comp.html#ss%3Awarn57

**Review by WardBrian — APPROVED** [review]
LGTM, thanks for hopping on this so quick

**Comment by SteveBronder** [comment]
Np!

**Pull request merged**